### PR TITLE
Refactoring Druid & SQLa into a proper "Connector" interface

### PR DESCRIPTION
**Pull request by mistercrunch**
Note that this will make it easier to build in new connectors, but may break existing ones in the wild.

### CODE REVIEW

File: superset/connectors/base.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+from sqlalchemy import Column, Integer, String, Text, Boolean
+
+from superset import utils
+from superset.models.helpers import AuditMixinNullable, ImportMixin
+
+
+class BaseDatasource(AuditMixinNullable, ImportMixin):
+
+    """A common interface to objects that are queryable (tables and datasources)"""
```
Comment (bkyryliuk) [inline superset/connectors/base.py]: Line too long (83 > 79 characters)

File: superset/connectors/base.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+    @property
+    def explore_url(self):
+        if self.default_endpoint:
+            return self.default_endpoint
+        else:
+            return "/superset/explore/{obj.type}/{obj.id}/".format(obj=self)
+
+    @property
+    def column_formats(self):
```
Comment (bkyryliuk) [inline superset/connectors/base.py]: have u meant metric_formats ?

File: superset/connectors/base.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        return {
+            m.metric_name: m.d3format
+            for m in self.metrics
+            if m.d3format
+        }
+
+    @property
+    def data(self):
+        """data representation of the datasource sent to the frontend"""
```
Comment (bkyryliuk) [inline superset/connectors/base.py]: s/data/Data

File: superset/connectors/base.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+            'filter_select': self.filter_select_enabled,
+            'filterable_cols': utils.choicify(self.filterable_column_names),
+            'gb_cols': utils.choicify(self.groupby_column_names),
+            'id': self.id,
+            'metrics_combo': self.metrics_combo,
+            'name': self.name,
+            'order_by_choices': order_by_choices,
+            'type': self.type,
+        }
+        if self.type == 'table':
```
Comment (bkyryliuk) [inline superset/connectors/base.py]: I would do this more like not to have if self.type = 'table' clause
```
d['granularity_sqla'] = self.granularity_sqla if self.granularity_sqla:
d['time_grain_sqla'] = self.time_grain_sqla if self.time_grain_sqla:
```

File: superset/connectors/druid/models.py [hunk 5]
```diff
@@ -0,0 +1,1062 @@
...
+            metric_type='count',
+            json=json.dumps({'type': 'count', 'name': 'count'})
+        ))
+        # Somehow we need to reassign this for UDAFs
+        if self.type in ('DOUBLE', 'FLOAT'):
+            corrected_type = 'DOUBLE'
+        else:
+            corrected_type = self.type
+
+        if self.sum and self.is_num:
```
Comment (bkyryliuk) [inline superset/connectors/druid/models.py]: could u please move this code to the function:
```
def get_druid_metric(self, name_str, col_name, corrected_type):
            mt = corrected_type.lower() + 'name_str'
            name = name_str + '__' + self.column_name
            metrics.append(DruidMetric(
                metric_name=name,
                metric_type=name_str.lower(),
                verbose_name='{}({})'.format(name_str.upper(), self.column_name),
                json=json.dumps({
                    'type': mt, 'name': name, 'fieldName': self.column_name})
           ))
```

File: superset/connectors/druid/models.py [hunk 6]
```diff
@@ -0,0 +1,1062 @@
...
+                .first()
+            )
+            metric.datasource_name = self.datasource_name
+            if not m:
+                new_metrics.append(metric)
+                session.add(metric)
+                session.flush()
+
+    @classmethod
+    def import_obj(cls, i_column):
```
Comment (bkyryliuk) [inline superset/connectors/druid/models.py]: s/i_column/i_metric

File: superset/connectors/druid/models.py [hunk 7]
```diff
@@ -0,0 +1,1062 @@
...
+        def lookup_obj(lookup_column):
+            return db.session.query(DruidColumn).filter(
```
Comment (bkyryliuk) [inline superset/connectors/druid/models.py]: s/DruidColumn/DruidMetric

**Comment by bkyryliuk** [comment]
Fun PR :) Great Job ! LGTM

File: superset/connectors/base.py (hunk 1 above)
Comment (mistercrunch) [inline superset/connectors/base.py]: Personally I'm ok with up to 85-90 and that's the way I set up pylint / landscape but I can't do it on codeclimate...

File: superset/connectors/base.py (hunk 2 above)
Comment (mistercrunch) [inline superset/connectors/base.py]: yeah I agree that this is weird but here I just moved the class without changing it...

File: superset/connectors/base.py (hunk 4 above)
Comment (mistercrunch) [inline superset/connectors/base.py]: We should use inheritance here and move this block to SqlaTable.data, I'll add a TODO

**Pull request merged**